File: boj/b_14502.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline

n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
from collections import deque
from copy import deepcopy
deq, result = deque(), 0
for i in range(n):
    for j in range(m):
        if arr[i][j] == 2:
            deq.append((i, j))
def bfs():
    q, array = deq.copy(), deepcopy(arr)
    while q:
        r, c = q.popleft()
        for d in ((-1, 0), (1, 0), (0, -1), (0, 1)):
            dr, dc = r + d[0], c + d[1]
            if dr < 0 or dc < 0 or dr >= n or dc >= m:
                continue
            if not array[dr][dc]:
                array[dr][dc] = 2
                q.append((dr, dc))
    res = 0
    for i in range(n):
        for j in range(m):
            if not array[i][j]:
                res += 1
    return res
def set_wall(cnt):
    global result
    if cnt == 3:
        result = max(result, bfs())
        return
    for i in range(n):
        for j in range(m):
            if not arr[i][j]:
                arr[i][j] = 1
                set_wall(cnt + 1)
                arr[i][j] = 0
set_wall(0)
print(result)

File: boj/b_14502.js
function solution(n, m, array) {
  const deq = [];
  for (let i = 0; i < n; i += 1) {
    for (let j = 0; j < m; j += 1) {
      if (array[i][j] === 2) deq.push([i, j]);
    }
  }
  const bfs = () => {
    const q = deq.slice();
    let front = 0;
    const arr = JSON.parse(JSON.stringify(array));
    while (q[front]) {
      const [r, c] = q[front++];
      for (const d of [
        [-1, 0],
        [1, 0],
        [0, -1],
        [0, 1],
      ]) {
        const [dr, dc] = [r + d[0], c + d[1]];
        if (dr < 0 || dc < 0 || dr >= n || dc >= m) continue;
        if (!arr[dr][dc]) {
          arr[dr][dc] = 2;
          q.push([dr, dc]);
        }
      }
    }
    let res = 0;
    for (let i = 0; i < n; i += 1) {
      for (let j = 0; j < m; j += 1) {
        if (!arr[i][j]) res += 1;
      }
    }
    return res;
  };
  let result = 0;
  const setWall = (cnt) => {
    if (cnt == 3) return (result = Math.max(result, bfs()));
    for (let i = 0; i < n; i += 1) {
      for (let j = 0; j < m; j += 1) {
        if (!array[i][j]) {
          array[i][j] = 1;
          setWall(cnt + 1);
          array[i][j] = 0;
        }
      }
    }
  };
  setWall(0);
  return result;
}
console.log(
  solution(7, 7, [
    [2, 0, 0, 0, 1, 1, 0],
    [0, 0, 1, 0, 1, 2, 0],
    [0, 1, 1, 0, 1, 0, 0],
    [0, 1, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 1],
    [0, 1, 0, 0, 0, 0, 0],
    [0, 1, 0, 0, 0, 0, 0],
  ])
);
